sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -1306,112 +1315,206 @@ public void populateDisplayData(DisplayData.Builder builder) {
       return withCreatWatermarkEstimatorFn(
           state -> {
             return new Manual(state);
           });
     }
 
-    // If a transactional producer is used and it's desired to only read records from committed
-    // transaction, it's recommended to set read_committed. Otherwise, read_uncommitted is the
-    // default value.
+    /**
+     * Sets "isolation_level" to "read_committed" in Kafka consumer configuration. This is ensures
+     * that the consumer does not read uncommitted messages. Kafka version 0.11 introduced
+     * transactional writes. Applications requiring end-to-end exactly-once semantics should only
+     * read committed messages. See JavaDoc for {@link KafkaConsumer} for more description.
+     */
     public ReadSourceDescriptors<K, V> withReadCommitted() {
       return withConsumerConfigUpdates(ImmutableMap.of("isolation.level", "read_committed"));
     }
 
+    /**
+     * Enable committing record offset. If {@link #withReadCommitted()} or {@link
+     * ConsumerConfig#ENABLE_AUTO_COMMIT_CONFIG} is set together with {@link #commitOffsets()},
+     * {@link #commitOffsets()} will be ignored.
+     */
     public ReadSourceDescriptors<K, V> commitOffsets() {
       return toBuilder().setCommitOffsetEnabled(true).build();
     }
 
+    /**
+     * Set additional configuration for the backend offset consumer. It may be required for a
+     * secured Kafka cluster, especially when you see similar WARN log message 'exception while
+     * fetching latest offset for partition {}. will be retried'.
+     *
+     * <p>In {@link ReadFromKafkaDoFn}, there're two consumers running in the backend actually:<br>
+     * 1. the main consumer, which reads data from kafka;<br>
+     * 2. the secondary offset consumer, which is used to estimate backlog, by fetching latest
+     * offset;<br>
+     *
+     * <p>By default, offset consumer inherits the configuration from main consumer, with an
+     * auto-generated {@link ConsumerConfig#GROUP_ID_CONFIG}. This may not work in a secured Kafka
+     * which requires more configurations.
+     */
     public ReadSourceDescriptors<K, V> withOffsetConsumerConfigOverrides(
         Map<String, Object> offsetConsumerConfig) {
       return toBuilder().setOffsetConsumerConfig(offsetConsumerConfig).build();
     }
 
+    /**
+     * Update configuration for the backend main consumer. Note that the default consumer properties
+     * will not be completely overridden. This method only updates the value which has the same key.
+     *
+     * <p>In {@link ReadFromKafkaDoFn}, there're two consumers running in the backend actually:<br>
+     * 1. the main consumer, which reads data from kafka;<br>
+     * 2. the secondary offset consumer, which is used to estimate backlog, by fetching latest
+     * offset;<br>
+     *
+     * <p>By default, main consumer uses the configuration from {@link
+     * KafkaIOUtils#DEFAULT_CONSUMER_PROPERTIES}.
```

Comment:
```suggestion
     * Replaces the configuration for the main consumer.
     *
     * <p>In {@link ReadFromKafkaDoFn}, there are two consumers running in the backend:
     * <ol>
     *   <li>the main consumer which reads data from kafka.
     *   <li>the secondary offset consumer which is used to estimate the backlog by fetching the latest offset.
     * </ol>
     *
     * <p>By default, main consumer uses the configuration from {@link
     * KafkaIOUtils#DEFAULT_CONSUMER_PROPERTIES}.
     *
     * See {@link #withConsumerConfigUpdates} for updating the configuration instead of overriding it.
```